### ISSUE TITLE
Deploying next.js app to vercel or netlify throws error

### ISSUE THREAD

**Issue by simonpeters** [nrwl/nx#5660]
## Current Behavior
Unable to deploy a fresh nx/next. js repo to vercel or netlify

## Expected Behavior
Be able to run the build command as described in the docs

## Steps to Reproduce
1. create new nx repo with the default next.js app
2. deploy to vercel as decribed  [here](https://nx.dev/latest/react/guides/nextjs#deploying-to-vercel)
3. have build fail

## Vercel

09:15:59.771 | >  NX   ERROR  Running target "check-in:build" failed
-- | --
09:15:59.771 | Failed tasks:
09:15:59.771 |  
09:15:59.771 | - check-in:build:production
09:15:59.789 | Error: Command "npx nx build check-in --prod --outputPath=." exited with 1
## Netlify
`10:29:49 PM: ────────────────────────────────────────────────────────────────
10:29:49 PM:   "build.command" failed                                        
10:29:49 PM: ────────────────────────────────────────────────────────────────
10:29:49 PM: ​
10:29:49 PM:   Error message
10:29:49 PM:   Command failed with exit code 1: nx build check-in --output-path=. --verbose
10:29:49 PM: ​
10:29:49 PM:   Error location
10:29:49 PM:   In Build command from Netlify app:
10:29:49 PM:   nx build check-in --output-path=. --verbose`

### Environment
This happens on netlify and on vercel

Link to [repo](https://github.com/SpacedBe/spaced-mono)

**Comment by simonpeters** [comment]
<img width="644" alt="Screenshot 2021-05-14 at 09 23 36" src="https://user-images.githubusercontent.com/714733/118236226-1aa44780-b496-11eb-98f3-ccd4d81872c6.png">

<img width="916" alt="Screenshot 2021-05-14 at 09 24 29" src="https://user-images.githubusercontent.com/714733/118236314-34458f00-b496-11eb-9384-ad7ae437e319.png">


Here also are the Vercel configuration and netlify configuration

**Comment by gregoryforel** [comment]
I think the **Publish directory** in Netlify should be `dist/apps/<your app name>`

**Comment by JamesHenry** [comment]
Confirmed, even locally running with `--outputPath=.` no longer works at all.

Bizarrely it doesn't seem to be down to the nextjs executor...

If I patch `build.impl.js` to noop and always resolve successfully:

```js
function buildExecutor(options, context) {
    return Promise.resolve({ success: true });
}
```

It still ends with a failure.

Any nested value within the workspace for `--outputPath`, e.g. `--outputPath=./tmp` does not have the same issue, but is not applicable as a solution for deploying to Vercel or Netlify so this needs to be addressed.

**Comment by JamesHenry** [comment]
I dug into this and have confirmed that the issue lies somewhere in the nx-cloud tasks runner, so I have updated the scope and assigned to Victor as I cannot investigate any further.

@simonpeters I have come up with a workaround to unblock deployments until this is resolved:

Go to your `nx.json` and add a new tasks runner entry within `tasksRunnerOptions` under your existing one (which will be called "default"), e.g. I have called mine `no-cloud`:

```diff
{
  // ...
  "tasksRunnerOptions": {
    "default": {
      "runner": "@nrwl/nx-cloud",
      "options": {
        "accessToken": "FOOBAR",
        "cacheableOperations": ["build", "test", "lint", "e2e"],
        "canTrackAnalytics": false,
        "showUsageWarnings": true
      }
    },
+   "no-cloud": {
+     "runner": "@nrwl/workspace/src/tasks-runner/default-tasks-runner",
+     "options": {}
+   }
  },
  // ...
}
```

Then append `--runner=no-cloud` to your build command that you run on Vercel or Netlify, e.g.

```sh
npx nx build your-project --outputPath=. --runner=no-cloud
```

This works for me locally and on Vercel.

**Comment by simonpeters** [comment]
@JamesHenry that works! Thanks so much for figuring out the workaround.

Cheers

**Comment by cyrus-za** [comment]
Thanks @JamesHenry that works for me too!

Question: Only 1 of my 4 next apps are running into this issue (after upgrading from 12.1 to 12.3). Why would this be? Also, why is there no error message? It literally finishes the nextjs build and then nx job fails without an error.

**Comment by vitorcamachoo** [comment]
Same issue here

**Comment by Bodom78** [comment]
While https://github.com/nrwl/nx/issues/5660#issuecomment-841639368 initially worked, I'm once again getting build failures on Vercel on NX 12.3.6. Anyone else having issues again with the no-cloud fix?


21:52:13.890 | ———————————————————————————————————————————————
-- | --
21:52:13.891 | &gt;  NX   ERROR  Running target "chicmanagement:build" failed
21:52:13.891 | Failed tasks:
21:52:13.891 | &nbsp;
21:52:13.891 | - chicmanagement:build:production
21:52:13.912 | Error: Command "npx nx build chicmanagement --prod --outputPath=. --runner=no-cloud" exited with 1

**Comment by LiamMyles** [comment]
Thanks, @JamesHenry https://github.com/nrwl/nx/issues/5660#issuecomment-841639368 was really helpful to get my app deployed.

Unfortunately, It seems to be missing an edge case for me. Not sure if this is the right place to raise it, but figured there is no harm in mentioning it for others :) 

In short, when `nx build your-project ` runs with the no-cloud runner it seems to miss files that should dynamically be added to nextjs's public folder.  (Note images seem to be fine, but they go into the a .next/cache folder it seems)

In my case, I have a component that imports fonts via styled-components that seem to work fine with the standard `nx build your-project` but don't get include when I build with `npx nx build your-project --outputPath=. --runner=no-cloud`

My clunky fix for this was just to add the fonts directly to the public folder of the next.js app using them. Which seems to keep both `nx build your-project ` and `--runner=no-cloud` happy.

Ideally, there is a workaround to get this to work without duplicating the fonts, but I'm not experienced enough with how NX runners work to fully understand what's going on here. So any guidance would be great!

**Comment by kirjai** [comment]
We have merged an update to the documentation. we no longer advise to use `--outputPath=.`. please use a custom build output path instead. Refer to the [updated documentation for more details](https://nx.dev/latest/react/guides/nextjs#existing-vercel-project).

If you experience any issues with the updated approach, please open a new issue

**Comment by JamesHenry** [comment]
@kirjai Did you try this with Nx Cloud caching involved?

Using .next as the output directory always used to cause problems, which is why we recommended the outputPath approach but I have applied it per the updated guidance and it works _as long as there isn't a cloud cache hit_. If the build artifacts are retrieved from the cache, Vercel will consistently complain of missing files (tested across two different Next.js projects):

![Untitled](https://user-images.githubusercontent.com/900523/122636194-ed803000-d0f8-11eb-820f-be67a91f497f.png)

I even tried changing things again to cause a cache miss (worked on Vercel), then reverting so it was a cache hit, the behaviour is consistent.

Please let me know if you can reproduce

**Comment by JamesHenry** [comment]
I'm pleased to report that after quite a bit of iteration it seems that my specific cache for my specific projects had ended up in a bad state. Unfortunately I cannot be 100% sure how that happened, but if anybody else finds themselves in this position after changing their Vercel config, I recommend you just create a commit which will bust the cache (i.e. make your project "affected" as far as Nx is concerned) for your particular project and push up to trigger a new build.

Since doing that I have not had any further issues so we don't need to open a new issue for this.

**Comment by kirjai** [comment]
regarding the error in https://github.com/nrwl/nx/issues/5660#issuecomment-864375021 , hopefully #6204 should help, once it's released.
it seems that the issue was affecting those that have been using Nx Cloud in conjunction with Vercel's build caching

**Comment by cyrus-za** [comment]
@JamesHenry @kirjai I am running into this now with one of my apps and busting the cache did not fix it. I tried to switch to using the custom output path in vercel config and removing the --outputPath=. but that makes me run into the build manifest issue. 

The documentaton linked in https://github.com/nrwl/nx/issues/5660#issuecomment-864221945 also does not make it clear what workspace.json should look like. So I created a new workspace and ran the commands just to see what it generates and there is an outputPath and outputs option in there still, so how does that work? if outputs for the cache is equal to the --outputPath then does that mean the cache will look at the default directory in the architect, despite vercel looking for it in a different directory?
Because the docs show screenshots on vercel where its in .next subdirectory and workspace shows the parent directory.

**Comment by cyrus-za** [comment]
Update: I left workspace.json as is and it now works after setting the output directory. I had to add --skip-nx-cache once before removing it to reset the cloud cache

**Comment by dylanwatsonsoftware** [comment]
Instead of adding the the no-cloud runner, I've just removed `build` as a cache able operation in the nx.json file. That way the other things are still cached.

**Comment by JamesHenry** [comment]
@dylanwatsonsoftware @cyrus-za No workarounds should be required any more. Please update to the latest Nx and if you encounter any problems please open a new issue with a minimal reproduction. Many thanks!

**Comment by dylanwatsonsoftware** [comment]
Ok! Interesting. Will give it a go. Thanks!

I thought I had but I'll go check.

**Comment by carlosfaria94** [comment]
We have followed this guide: https://nx.dev/recipes/other/deploy-nextjs-to-vercel but the `public` folder is not being considered.

Steps:
- Create a new app: `npx nx g @nrwl/next:app my-new-app`
- Add a `robots.txt` inside of `apps/my-new-app/public`
- Run `npx nx build my-new-app --prod` 
- `robots.txt` is not included in `dist/apps/my-new-app/.next` 🚨

Result: no `robots.txt` for our website in Vercel

The only solution that worked for us is to add `--outputPath=.` to the Vercel build command.

Is this still an issue?

--- 

I don't know if it's related, but we cannot use Nx Cloud Cache in Vercel, because when we get a hit in Vercel the files are not found:

![image](https://user-images.githubusercontent.com/4710889/208424383-bb79bfc1-a82c-4732-8e53-da27acb777ae.png)

I have set an env variable `NX_CACHE_DIRECTORY=/tmp/nx-cache`

Nx 14.8.6
Next 12.3.1
